llvm/include/llvm/CAS/CASDB.h

```diff
@@ -392,6 +392,10 @@ class HierarchicalTreeBuilder {
     return pushImpl(ID, Kind, Path);
   }
 
+  /// Add a directory with specific contents. Allows merging the contents of
+  /// multiple directories.
+  Error pushDirectory(CASDB &CAS, CASID ID, const Twine &Path);
```

Comment:
I suggest a different name since it behaves quite differently.
- `push()` takes an ID and path and makes the path get exactly that ID.
- `pushDirectory()` takes a path and makes it exist.
-  this new function takes an ID and a path and merges it with anything else that comes in

Maybe `pushTreeContent()`?

I think the documentation should be specific about behaviour. It should be equivalent to:
- Calling `pushDirectory()` for every tree
- Calling `push()` for every non-tree

In terms of API, please skip the `CASDB` here and instead modify `HierarchicalTreeEntry` to carry a bit:
```
  private:
    Optional<CASID> ID;
    TreeEntry::EntryKind Kind;
    bool MergeContentRecursively = false; // true for pushTreeContents
    std::string Path;
```
Then the lookups can be lazy (and eventually optimized when there's nothing to merge).